config/internal/configsource/component/configsource.go

```diff
@@ -0,0 +1,89 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package component
+
+import (
+	"context"
+
+	"go.uber.org/zap"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config/internal/configsource/configmodels"
+)
+
+// ConfigSource is the interface to be implemented by objects used by the collector
+// to retrieve external configuration information.
+type ConfigSource interface {
+	component.Component
+
+	// BeginSession signals that the ConfigSource is about to be used to inject data into
+	// the configuration. ConfigSource objects can assume that there won't be
+	// concurrent sessions and can use this call according to their needs:
+	// lock needed resources, suspend background tasks, etc.
+	BeginSession(ctx context.Context) error
+
+	// Apply retrieves generic values given the arguments specified on a specific
```

Comment:
Perhaps I need some coffee, but I couldn't grasp it without looking at the example. Here's my suggestion, assuming that my understanding is correct:

```
Apply goes to the configuration source, sending the given parameters as a map and returns the resulting configuration value or snippet. A configuration snippet will become a map, as follows:

	//  $my_config_src:
	//    param0: true
	//    param1: "some string"

Becomes a call with the following payload as params:

	//  map[string]interface{}{
	//    "param0": true,
	//    "param1": "some string",
 	// }

A ConfigSource should then unmarshal or cast the params. An error should be returned when the params don't fit the expected usage.
```

Reply:
This is probably stretching it but do we expect different config sources to have wildly different params? Can we may be arrive at a common set of params for all config sources? For example, `lookup_key` + `namespace` would probably satisfy most config sources. If we need to add more later, we can add more param fields while keeping all fields nullable. This would be better in two ways:

1. Consistency for users across config sources. Users will not have to learn a new set of params for each source. Inconsistency here can cause a lot of pain for end users IMO. 
2. We pass a typed params object to all the Apply method.

Somewhat unrelated but in addition to this, can we have config sources return the snippet that would replace the it's placeholder in config as a byte slice or YAML object? This would simplify the Apply method (we could actually rename it to Get()`) and shift even more logic/complexity to config loading code in core. As a result writing sources would become a lot simpler and obvious I think. May be if sources are only responsible for returning raw values, their implementations don't need to worry about much else either. Core can worry about concurrency and locks.

I feel the current design is accounting for too much flexibility and could be simplified by not supporting some use cases but if those use cases are very real and not hypothetical then I take my words back :)

Reply:
@jpkrohling perhaps I didn't have enough coffee before writing it? Incorporated your suggestion.

@owais let me know if there was something that we didn't cover in our chat today.